core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java

```diff
@@ -85,31 +87,40 @@ public String toString()
 
     public static class Accumulator
     {
-        private final Map<String, Metric<?>> merged = new HashMap<>();
+        private final Map<String, List<Metric<?>>> groupedMetrics = new HashMap<>();
 
         private Accumulator()
         {
         }
 
-        public Accumulator add(Metrics metrics)
+        public Accumulator add(List<Metrics> metricsList)
         {
-            metrics.getMetrics().forEach((key, value) ->
-                    merged.merge(key, value, Accumulator::merge));
+            metricsList.forEach(this::add);
             return this;
         }
 
-        @SuppressWarnings({"rawtypes", "unchecked"})
-        private static Metric<?> merge(Metric<?> a, Metric<?> b)
+        public Accumulator add(Metrics metrics)
         {
-            return (Metric<?>) ((Metric) a).mergeWith(b);
+            metrics.getMetrics().forEach((key, value) ->
+                    groupedMetrics.computeIfAbsent(key, ignored -> new ArrayList<>()).add(value));
+            return this;
         }
 
         public Metrics get()
         {
-            if (merged.isEmpty()) {
+            if (groupedMetrics.isEmpty()) {
                 return EMPTY;
             }
+
+            Map<String, Metric<?>> merged = new HashMap<>();
+            groupedMetrics.forEach((key, values) -> merged.put(key, merge(values.get(0), values.subList(1, values.size()))));
```

Comment:
Is there a change that there will be a single metric key where we will have only one value?

Reply:
Yes. In case of a single task.